File: ts/src/parse.test.ts
import { parse } from "./parse";
import * as fs from "fs";
import { basename, extname, join } from "path";

describe("parse", () => {
  const testDataDir = join(__dirname, "../../testdata/parse");
  const testFiles = fs
    .readdirSync(testDataDir)
    .filter((file) => file.endsWith(".prompt.md"))
    .map((file) => ({
      name: basename(file, extname(file)),
      file,
    }));

  it.each(testFiles)("$name", ({ name, file }) => {
    const promptFile = join(testDataDir, file);
    const expectedFile = join(
      testDataDir,
      file.replace(".prompt.md", ".expected.json")
    );

    const promptContent = fs.readFileSync(promptFile, "utf-8");
    const expectedContent = fs.readFileSync(expectedFile, "utf-8");
    const expected = JSON.parse(expectedContent);

    const result = parse(promptContent);
    expect(result).toEqual(expected);
  });
});

File: go/parse_test.go
package promptmd

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestParse(t *testing.T) {
	testDataDir := filepath.Join("..", "testdata", "parse") // Assuming testdata is one level up
	testFiles, err := os.ReadDir(testDataDir)
	if err != nil {
		t.Fatalf("Failed to read testdata directory: %v", err)
	}

	var promptFiles []map[string]string

	for _, file := range testFiles {
		if strings.HasSuffix(file.Name(), ".prompt.md") {
			name := strings.TrimSuffix(file.Name(), filepath.Ext(file.Name()))
			promptFiles = append(promptFiles, map[string]string{
				"name": name,
				"file": file.Name(),
			})
		}
	}

	for _, testFile := range promptFiles {
		name := testFile["name"]
		file := testFile["file"]

		t.Run(name, func(t *testing.T) {
			promptFile := filepath.Join(testDataDir, file)
			expectedFile := filepath.Join(testDataDir, strings.Replace(file, ".prompt.md", ".expected.json", 1))

			promptContentBytes, err := os.ReadFile(promptFile)
			if err != nil {
				t.Fatalf("Failed to read prompt file: %v", err)
			}
			promptContent := string(promptContentBytes)

			expectedContentBytes, err := os.ReadFile(expectedFile)
			if err != nil {
				t.Fatalf("Failed to read expected file: %v", err)
			}
			expectedContent := string(expectedContentBytes)

			var expected *Prompt
			err = json.Unmarshal([]byte(expectedContent), &expected)
			if err != nil {
				t.Fatalf("Failed to unmarshal expected JSON: %v", err)
			}

			result, err := Parse(promptContent)

			require.NoError(t, err)
			require.Equal(t, expected, result)
		})
	}
}
